mopidy/files/library.py

```diff
@@ -0,0 +1,157 @@
+            for name in os.listdir(localpath):
+                child = os.path.join(localpath, name)
+                uri = path.path_to_uri(child)
+                name = name.decode('ascii', 'ignore')
+                if self._follow_symlinks:
+                    st = os.stat(child)
+                else:
+                    st = os.lstat(child)
+                if not self._show_hidden and name.startswith(b'.'):
+                    continue
+                elif stat.S_ISDIR(st.st_mode):
+                    result.append(models.Ref.directory(name=name, uri=uri))
+                elif stat.S_ISREG(st.st_mode) and self._check_audiofile(uri):
+                    result.append(models.Ref.track(name=name, uri=uri))
+                else:
+                    logger.warn(u'Ignored file: %s' % child.decode('ascii',
+                                                                   'ignore'))
+                    continue
+
+        result.sort(key=operator.attrgetter('name'))
+        return result
+
+    def lookup(self, uri):
+        logger.debug(u'looking up uri = %s' % uri)
+        localpath = path.uri_to_path(uri)
+        if not self._is_in_basedir(localpath):
+            logger.warn(u'Not in basedir: %s' % localpath)
+            return []
+        try:
+            result = self._scanner.scan(uri)
+            track = utils.convert_tags_to_track(result.tags).copy(
+                uri=uri, length=result.duration)
+        except exceptions.ScannerError as e:
+            logger.warning(u'Problem looking up %s: %s', uri, e)
+            track = models.Track(uri=uri)
+            pass
+        if not track.name:
+            filename = os.path.basename(localpath)
+            name = urllib2.unquote(filename).decode('ascii', 'ignore')
+            track = track.copy(name=name)
+        return [track]
+
+    def _show_media_dirs(self):
+        result = []
+        for media_dir in self._media_dirs:
+            dir = models.Ref.directory(
+                name=media_dir['name'],
+                uri=path.path_to_uri(media_dir['path']))
+            result.append(dir)
+        return result
+
+    def _check_audiofile(self, uri):
+        try:
+            result = self._scanner.scan(uri)
+            logger.debug(u'got scan result playable: %s for %s' %
+                         (result.uri, str(result.playable)))
+            res = result.playable
+        except exceptions.ScannerError as e:
+            logger.warning(u'Problem scanning %s: %s', uri, e)
+            res = False
+        return res
+
+    def _is_playlist(self, child):
+        return os.path.splitext(child)[1] == '.m3u'
+
+    def _is_in_basedir(self, localpath):
+        res = False
+        basedirs = [mdir['path'] for mdir in self._media_dirs]
+        for basedir in basedirs:
+            if basedir == localpath:
+                res = True
+            else:
+                try:
+                    path.check_file_path_is_inside_base_dir(localpath, basedir)
```

Comment:
Hm, i don't like the user to browse outside the media directory seeing a lot of empty directories.

    def is_local_path_inside_base_dir(local_path, base_path):
        if local_path.endswith(os.sep):
            raise ValueError('Local path %s cannot end with a path separator'
                           % local_path)
        # Expand symlinks
        real_base_path = os.path.realpath(base_path)
        real_local_path = os.path.realpath(local_path)

        if os.path.isfile(local_path):
            # Use dir of file for prefix comparision, so we don't accept
            # /tmp/foo.m3u as being inside /tmp/foo, simply because they have a
            # common prefix, /tmp/foo, which matches the base path, /tmp/foo.
            real_local_path = os.path.dirname(real_local_path)

        # Check if dir of file is the base path or a subdir
        common_prefix = os.path.commonprefix([real_base_path, real_local_path])
        return common_prefix == real_base_path

This would check on files and directories. I do not understand why the check on endswith(os.sep) 
Would path.realpath() not tke care of that case?